### PR TITLE
iOS - Sync 1.16.x - Uplift

### PR DESCRIPTION
**Pull request by Brandon-T**
Uplift of #6722 for 1.16.x

- Adds support for iOS Sync chain with Brave Desktop.
- Adds support for iOS Chromium bookmarks so we can sync them with Desktop.
- Adds importing and exporting Bookmarks to HTML for iOS.
- Adds better compatibility with iOS and Chromium core in general.
- All interfaces that are native to iOS must be explicitly exported with OBJC_EXPORT
- Rewards database code runs on Chromium Threads instead of iOS DispatchQueues.

## Submitter Checklist:

- [ ] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [ ] Verified that all lint errors/warnings are resolved (`npm run lint`, `npm run gn_check`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Requested a security/privacy review as needed.
- [ ] Added appropriate QA labels (`QA/Yes` or `QA/No`) to the associated issue
- [ ] Added appropriate release note labels (`release-notes/include` or `release-notes/exclude`) to the associated issue
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

**Comment by kjozwiak** [comment]
Restarting `macOS` via https://ci.brave.com/job/pr-brave-browser-ios-sync-1.16.x-macos/6/ as `build` failed as per https://ci.brave.com/job/pr-brave-browser-ios-sync-1.16.x-macos/5/execution/node/202/log/

```
02:05:29  FAILED: obj/content/browser/browser/browser_main_loop.o 
```

Restarting `iOS` via https://ci.brave.com/job/pr-brave-browser-ios-sync-1.16.x-ios/7/ due to `build` failing as per https://ci.brave.com/job/pr-brave-browser-ios-sync-1.16.x-ios/6/execution/node/185/log/

```
01:28:05  FAILED: gen/brave/vendor/brave-ios/ledger.mojom.objc.h gen/brave/vendor/brave-ios/ledger.mojom.objc+private.h gen/brave/vendor/brave-ios/ledger.mojom.objc.mm gen/brave/vendor/brave-ios/objc_templates_bytecode/ledger_objc_templates.zip 
```

@bsclifton @Brandon-T just double checking, are we fine with the following `lint` issues or should these be addressed?

* https://ci.brave.com/job/pr-brave-browser-ios-sync-1.16.x-android/5/execution/node/188/log/
* https://ci.brave.com/job/pr-brave-browser-ios-sync-1.16.x-linux/5/execution/node/180/log/

**Comment by bsclifton** [comment]
Above errors look to be intermittent; just rebased after latest Chromium bump so that we can get a clean run 👍

**Comment by bsclifton** [comment]
Marked as `PR/blocked` since we need to merge https://github.com/brave/brave-core/pull/6722 first

Will ping here and remove label once things are in better shape 😄

**Comment by bsclifton** [comment]
OK we're good to go on reviewing/approving/merging; gonna check out CI here and see what problems were flagged

**Comment by bsclifton** [comment]
@Brandon-T I'm seeing some errors here on the iOS run which may need addressing. The other platforms are OK

If you have Jenkins access, you can view the unit test error I am seeing here:
https://ci.brave.com/job/pr-brave-browser-ios-sync-1.16.x-ios/14/execution/node/213/log/

I think there might be a missing iOS guard around some of the rewards tests (which causes a linker error on `unit_test` target):
```
00:46:58  FAILED: obj/brave/vendor/brave-ios/tests/x64/brave_rewards_ios_tests 
00:46:58  TOOL_VERSION=1605485041 ../../build/toolchain/mac/linker_driver.py -Wcrl,strippath,/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/strip clang++ -B /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/  -Xlinker -rpath -Xlinker @executable_path/Frameworks -Xlinker -rpath -Xlinker @loader_path/Frameworks -Xlinker -rpath -Xlinker @executable_path/Frameworks -Xlinker -objc_abi_version -Xlinker 2 -Xlinker -sectcreate -Xlinker __TEXT -Xlinker __entitlements -Xlinker obj/brave/vendor/brave-ios/tests/brave_rewards_ios_tests.xcent -Wl,-fatal_warnings -arch x86_64 -Werror -Wl,-dead_strip -Wl,-no_data_in_code_info -Wl,-no_function_starts -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator13.6.sdk -mios-simulator-version-min=12.0 -Wl,-ObjC -Wl,-no_compact_unwind -F/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/Library/Frameworks -F. -o "obj/brave/vendor/brave-ios/tests/x64/brave_rewards_ios_tests" -Wl,-filelist,"obj/brave/vendor/brave-ios/tests/x64/brave_rewards_ios_tests.rsp" -framework CoreData -framework UIKit -framework CoreFoundation -framework CoreGraphics -framework CoreText -framework Foundation -framework XCTest -framework BraveRewards -framework MaterialComponents -framework Security -framework CFNetwork -framework MobileCoreServices -framework SystemConfiguration  brave_rust/out/x64/x86_64-apple-ios/release/libbrave_rust.a -lresolv
00:46:58  Undefined symbols for architecture x86_64:
00:46:58    "_OBJC_CLASS_$_RecurringDonation", referenced from:
00:46:58        objc-class-ref in ledger_database_test.o
00:46:58    "_OBJC_METACLASS_$_DataController", referenced from:
00:46:58        _OBJC_METACLASS_$_TempTestDataController in ledger_database_test.o
00:46:58    "_OBJC_CLASS_$_ServerPublisherInfo", referenced from:
00:46:58        objc-class-ref in ledger_database_test.o
00:46:58    "_OBJC_CLASS_$_MediaPublisherInfo", referenced from:
00:46:58        objc-class-ref in ledger_database_test.o
00:46:58    "_OBJC_CLASS_$_PendingContribution", referenced from:
00:46:58        objc-class-ref in ledger_database_test.o
00:46:58    "_OBJC_CLASS_$_Promotion", referenced from:
00:46:58        objc-class-ref in ledger_database_test.o
00:46:58    "_OBJC_CLASS_$_ContributionPublisher", referenced from:
00:46:58        objc-class-ref in ledger_database_test.o
00:46:58    "_OBJC_CLASS_$_ContributionInfo", referenced from:
00:46:58        objc-class-ref in ledger_database_test.o
00:46:58    "_OBJC_CLASS_$_ServerPublisherAmount", referenced from:
00:46:58        objc-class-ref in ledger_database_test.o
00:46:58    "_OBJC_CLASS_$_ActivityInfo", referenced from:
00:46:58        objc-class-ref in ledger_database_test.o
00:46:58    "_OBJC_CLASS_$_ContributionQueue", referenced from:
00:46:58        objc-class-ref in ledger_database_test.o
00:46:58    "_OBJC_CLASS_$_ServerPublisherLink", referenced from:
00:46:58        objc-class-ref in ledger_database_test.o
00:46:58    "_OBJC_CLASS_$_DataController", referenced from:
00:46:58        objc-class-ref in ledger_database_test.o
00:46:58        _OBJC_CLASS_$_TempTestDataController in ledger_database_test.o
00:46:58    "_OBJC_CLASS_$_UnblindedToken", referenced from:
00:46:58        objc-class-ref in ledger_database_test.o
00:46:58    "_OBJC_CLASS_$_PublisherInfo", referenced from:
00:46:58        objc-class-ref in ledger_database_test.o
00:46:58    "_OBJC_CLASS_$_ServerPublisherBanner", referenced from:
00:46:58        objc-class-ref in ledger_database_test.o
00:46:58    "_OBJC_CLASS_$_PromotionCredentials", referenced from:
00:46:58        objc-class-ref in ledger_database_test.o
00:46:58    "_OBJC_CLASS_$_BATLedgerDatabase", referenced from:
00:46:58        objc-class-ref in ledger_database_test.o
00:46:58  ld: symbol(s) not found for architecture x86_64
00:46:58  clang: error: linker command failed with exit code 1 (use -v to see invocation)
00:46:58  Traceback (most recent call last):
00:46:58    File "../../build/toolchain/mac/linker_driver.py", line 287, in <module>
00:46:58      Main(sys.argv)
00:46:58    File "../../build/toolchain/mac/linker_driver.py", line 97, in Main
00:46:58      subprocess.check_call(compiler_driver_args, env=env)
00:46:58    File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 190, in check_call
00:46:58      raise CalledProcessError(retcode, cmd)
00:46:58  subprocess.CalledProcessError: Command '['clang++', '-B', '/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/', '-Xlinker', '-rpath', '-Xlinker', '@executable_path/Frameworks', '-Xlinker', '-rpath', '-Xlinker', '@loader_path/Frameworks', '-Xlinker', '-rpath', '-Xlinker', '@executable_path/Frameworks', '-Xlinker', '-objc_abi_version', '-Xlinker', '2', '-Xlinker', '-sectcreate', '-Xlinker', '__TEXT', '-Xlinker', '__entitlements', '-Xlinker', 'obj/brave/vendor/brave-ios/tests/brave_rewards_ios_tests.xcent', '-Wl,-fatal_warnings', '-arch', 'x86_64', '-Werror', '-Wl,-dead_strip', '-Wl,-no_data_in_code_info', '-Wl,-no_function_starts', '-isysroot', '/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator13.6.sdk', '-mios-simulator-version-min=12.0', '-Wl,-ObjC', '-Wl,-no_compact_unwind', '-F/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/Library/Frameworks', '-F.', '-o', 'obj/brave/vendor/brave-ios/tests/x64/brave_rewards_ios_tests', '-Wl,-filelist,obj/brave/vendor/brave-ios/tests/x64/brave_rewards_ios_tests.rsp', '-framework', 'CoreData', '-framework', 'UIKit', '-framework', 'CoreFoundation', '-framework', 'CoreGraphics', '-framework', 'CoreText', '-framework', 'Foundation', '-framework', 'XCTest', '-framework', 'BraveRewards', '-framework', 'MaterialComponents', '-framework', 'Security', '-framework', 'CFNetwork', '-framework', 'MobileCoreServices', '-framework', 'SystemConfiguration', 'brave_rust/out/x64/x86_64-apple-ios/release/libbrave_rust.a', '-lresolv']' returned non-zero exit status 1
```

Besides this, there is also a GN check error which you can see here:
https://ci.brave.com/job/pr-brave-browser-ios-sync-1.16.x-ios/14/execution/node/198/log/

I think there are some missing GN deps. I'll look at fixing that one

**Comment by bsclifton** [comment]
Pushed fix for GN check; the unit test linker errors might be resolved by removing obsolete tests from `brave\vendor\brave-ios\tests\ledger_database_test.mm`
cc: @kylehickinson

**Comment by kylehickinson** [comment]
@bsclifton There are no obsolete tests in `ledger_database_test.mm` and all those files exist still in `brave-ios/Ledger/Data/Models`. Not sure why a linker error is coming up for those files though. Were they removed from GN?

File: ios/browser/api/bookmarks/BUILD.gn [hunk 1]
```diff
@@ -19,10 +19,12 @@ source_set("bookmarks") {
   deps = [
     "//base",
     "//components/bookmarks/browser",
+    "//components/prefs:prefs",
```
Comment (bridiver) [inline ios/browser/api/bookmarks/BUILD.gn:22]: `:prefs` is redundant. Same with many others in this change.

Comment (bsclifton) [reply ios/browser/api/bookmarks/BUILD.gn:22]: Pretty sure I added this after warnings were raised in gn check

**Comment by bsclifton** [comment]
All GN check errors resolved 🎉 Now just need to solve the iOS unit test failures:
https://ci.brave.com/job/pr-brave-browser-ios-sync-1.16.x-ios/17/execution/node/213/log/

**Comment by bsclifton** [comment]
Thanks for fixing those @Brandon-T 😄 

Windows ran into some environment issues w/ CI, but it has passed before. Will go ahead and merge! 😄  Let's make sure the lint / GN check /  `OBJC_EXPORT` changes are in the master PR too

**Review by bsclifton — APPROVED** [review]
++

**Pull request merged**